pkg/infra/kubernetes/exec_unit.go

```diff
@@ -30,6 +33,52 @@ type HelmExecUnit struct {
 	HorizontalPodAutoscaler *core.SourceFile
 }
 
+type manifestTransformer[K runtime.Object] struct {
+	fieldToTransform func(unit *HelmExecUnit) *core.SourceFile
+	transform        func(unit *HelmExecUnit, cfg config.ExecutionUnit, target K, log *zap.SugaredLogger) ([]HelmChartValue, error)
+	// readF is the function to read a given file to the runtime.Object that we will then cast down to K. You may leave
+	// it blank, in which case the transformer will assume readFile.
+	readF func(f *core.SourceFile) (runtime.Object, error)
+}
+
+func (transformer manifestTransformer[K]) apply(unit *HelmExecUnit, cfg config.ExecutionUnit) ([]HelmChartValue, error) {
+	source := transformer.fieldToTransform(unit)
+	log := zap.L().Sugar().With(logging.FileField(source), zap.String("unit", unit.Name))
+	log.Debugf("Transforming file, %s, for exec unit, %s", source.Path(), unit.Name)
+	readF := transformer.readF
+	if readF == nil {
+		readF = readFile
+	}
+	obj, err := readF(source)
+	if err != nil {
+		return nil, err
+	}
+	transformObj, ok := obj.(K)
+	if !ok {
+		t := reflect.TypeOf((*K)(nil))
+		for t.Kind() == reflect.Pointer {
+			t = t.Elem()
+		}
+		err = fmt.Errorf("expected file %s to contain %s Kind", source.Path(), t.Name())
```

Comment:
```suggestion
		var k K
		err = fmt.Errorf("expected file %s to contain %T Kind", source.Path(), t)
```
nit: a little more succinct

Reply:
Your version is a bit different, but probably good enough to warrant the succinctness. The output is:

```diff
- expected file test/templates/unit-deployment.yaml to contain Deployment Kind
+ expected file test/templates/unit-deployment.yaml to contain *v1.Deployment Kind
```